client/trino-jdbc/src/main/java/io/trino/jdbc/AsyncResultIterator.java

```diff
@@ -66,16 +67,23 @@ public class AsyncResultIterator
         this.finished = false;
         this.future = executorService.submit(() -> {
             try {
+                int rowsProcessed = 0;
                 do {
                     QueryStatusInfo results = client.currentStatusInfo();
                     progressCallback.accept(QueryStats.create(results.getId(), results.getStats()));
                     warningsManager.addWarnings(results.getWarnings());
                     for (List<Object> row : client.currentRows()) {
                         rowQueue.put(row);
-                        semaphore.release();
+                        if (rowsProcessed++ % BATCH_SIZE == 0) {
```

Comment:
You are emitting semaphore after 1st, 101st, 201st row?
Did you intent to go with 100th, 200th, 300th?